99/tc 499 implement update by query

emarcey opened the pull request:
Resolves issue #1050 

Creates object UpdateByQuery, with associated method in Index.

UpdateByQuery wraps the [_update_by_query](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-update-by-query.html) endpoint, which allows a search query & script to update a field in all matching docs.

Re-uses much of Search functionality (with fewer options, like no aggregations), and swaps Search's script_fields attribute with script that matches the scripting documented in the [elasticsearch API](https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-scripting-using.html)

HonzaKral commented on docs/update_by_query.rst:
```diff
@@ -0,0 +1,148 @@
+the ``Update By Query`` object APIs.
+
+You can pass an instance of the low-level `elasticsearch client <https://elasticsearch-py.readthedocs.io/>`_ when
+instantiating the ``Update By Query`` object:
```
can we instead point to the `configuration` chapter to promote global configuration instead of passing the client everywhere explicitly?

i know you mention it lower but that should be the recommended way and should be more highlighted.

HonzaKral commented on elasticsearch_dsl/response/__init__.py:
```diff
@@ -84,3 +84,16 @@ def __iter__(self):
+        super(UpdateByQueryResponse, self).__init__(response)
+
+    def __getitem__(self, attr_name):
+        return super(UpdateByQueryResponse, self).__getitem__(attr_name)
```
I don't think this method is needed.

HonzaKral commented on elasticsearch_dsl/update_by_query.py:
```diff
@@ -0,0 +1,149 @@
+
+        ubq._response_class = self._response_class
+        ubq._script = self._script.copy()
+        getattr(ubq, 'query')._proxied = getattr(self, 'query')._proxied
```
`ubq.query._proxied = self.query._proxied` would be more readable here

HonzaKral commented on elasticsearch_dsl/update_by_query.py:
```diff
@@ -0,0 +1,149 @@
+        Execute the search and return an instance of ``Response`` wrapping all
+        the data.
+        """
+        if ignore_cache or not hasattr(self, '_response'):
```
I don't think we should ever attempt to cache the response to update by query as it is not a read-only operation like `Search`

HonzaKral commented on test_elasticsearch_dsl/test_integration/test_update_by_query.py:
```diff
@@ -0,0 +1,35 @@
+    assert response.updated == 1
+    assert response.version_conflicts == 1
+
+def test_delete_by_query_with_script(data_client):
```
This looks like a dangerous test - `data_client` uses the `git` and `flat-git` indices in a read-only fashion and it is therefore not torn down after every single test. Altering any data in those indices will have consequences in other tests and is to be avoided. Please use `write_client` and an index starting with `test-` for any test that alters documents in elasticsearch.

HonzaKral commented on test_elasticsearch_dsl/test_update_by_query.py:
```diff
@@ -0,0 +1,286 @@
+from elasticsearch_dsl import UpdateByQuery, query, Q, Document
+
+
+# A lot of the basic query testing for Search can be copied over here and re-used
```
there is no point in copying unit tests when your class uses the parent methods, it just introduces duplications. If you are not reimplmenting methods like `index()` do not copy their tests here.

emarcey replied:
Thanks for your comments - I'm going to start working on them today.

But one question here: the UpdateByQuery doesn't inherit from Search because the query capabilities are so different - I can slim down the tests used here. Or would you rather I refactor to use UpdateByQuery inherits from Search to reduce duplication?

HonzaKral replied:
True, but some of the tests here (like `.index()`) test the functionality of `Request` that `UpdateByQuery` does extend. Also it might be ok to create another class below `Request` that both `Search` and `UpdateByQuery` can then extend that would contain the common functionality, what do you think?

HonzaKral replied:
Looking at it further it's just `query`, `filter`, and `exclude`. I would be ok with just a simple test each to verify they work since it's the same code as in `Search` where it is tested more extensively. Alternatively another class (either to extend or to mix in) is always a solution too.

emarcey commented:
@HonzaKral 

I think I addressed all of your comments. Let me know if there's something I missed.

Re: unit tests. For now, I slimmed down the tests to cover only what I have changed. I'm going to make a note to try and re-factor some time in the future

HonzaKral commented:
Thank you so much for the feature, @emarcey ! Much appreciated!

Sphinxs commented:
Does it support Python `dict` ? When I submit `script(source=f"ctx._source.test={dumps({'test': 1})}")` I receive: `TransportError(500, 'script_exception', 'compile error')`, but it works for strings, lists...
